File: offer/42/JPhoebe.py
class Solution:
    def FindNumbersWithSum(self, array, tsum):
        left = 0
        right = len(array) - 1

        while True:
            if left >= right:
                return []
            if array[left] + array[right] == tsum:
                return [array[left], array[right]]
            elif array[left] + array[right] > tsum:
                right -= 1
            elif array[left] + array[right] < tsum:
                left += 1
        # 2 + 8 / 4 + 6 / (2+x)(8-x) / 6x >= x*x
        # so 是第一个

File: offer/58/jphoebe.java
import java.util.concurrent.atomic.AtomicInteger;

public class Solution {

    boolean isSymmetrical(TreeNode pRoot) {
        if (pRoot == null){
            return true;
        }
        TreeNode left = pRoot.left;
        TreeNode right = pRoot.right;
        while (left != null && right != null){
            TreeNode leftTemp = left;
            TreeNode rightTemp = right;
            while (leftTemp != null && rightTemp != null){
                if (leftTemp.val != rightTemp.val){
                    return false;
                }
                leftTemp = leftTemp.left;
                rightTemp = rightTemp.right;
            }
            if (!this.compare(leftTemp, rightTemp)){
                return false;
            }
            left = left.right;
            right = right.left;
        }
        return this.compare(left, right);
    }

    public boolean compare(TreeNode left, TreeNode right){
        if ( (left==null && right!=null) || left!=null && right==null ){
            return false;
        }else{
            return true;
        }
    }

}
